kani-compiler/src/kani_middle/transform/dump_mir_pass.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Global transformation pass, which does not modify bodies but dumps MIR whenever the appropriate debug flag is passed.
+
+use crate::kani_middle::reachability::CallGraph;
+use crate::kani_middle::transform::GlobalPass;
+use crate::kani_queries::QueryDb;
+use kani_metadata::ArtifactType;
+use rustc_middle::ty::TyCtxt;
+use rustc_session::config::OutputType;
+use stable_mir::mir::mono::{Instance, MonoItem};
+use stable_mir::mir::Body;
+use std::fmt::{Debug, Formatter, Result};
+use std::fs::File;
+use std::io::BufWriter;
+use std::io::Write;
+
+/// Dump all MIR bodies.
+pub struct DumpMirPass {
+    enabled: bool,
+}
+
+impl DumpMirPass {
+    pub fn new(tcx: TyCtxt) -> Self {
+        Self { enabled: tcx.sess.opts.output_types.contains_key(&OutputType::Mir) }
+    }
+}
+
+impl Debug for DumpMirPass {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        f.debug_struct("DumpMirPass").field("enabled", &self.enabled).finish()
+    }
+}
+
+impl GlobalPass for DumpMirPass {
+    fn is_enabled(&self, _query_db: &QueryDb) -> bool {
+        self.enabled
+    }
+
+    fn transform(
+        &mut self,
+        tcx: TyCtxt,
+        starting_items: &[MonoItem],
+        bodies: Vec<(Body, Instance)>,
```

Comment:
I don't think we should be passing the vector of bodies. We could pass the `BodyTransformer` instead or some abstraction. It would be nice to be able to change this if the memory consumption gets too high.

Reply:
The problem with passing `&mut BodyTransformer` is that now there are two mutable references to it, the other one from the global pass we are mutably iterating over. I added a hacky solution to that by moving the global passes out before iteration and then moving them in, but I wonder whether it's worth it